tools/src/main/scala/scala/scalanative/codegen/llvm/CodeGen.scala

```diff
@@ -68,12 +68,40 @@ object CodeGen {
       val env = assembly.map(defn => defn.name -> defn).toMap
       val workDir = VirtualDirectory.real(config.workDir)
 
+      /**
+       *    - \C:a\b\c.txt -> a\b\c.txt
+       *    - C:a\b\c.txt -> a\b\c.txt
+       *    - p:\a\b\c.txt -> a\b\c.txt
+       *    - /a/b/c.txt -> a/b/c.txt
+       */
+      def dropPrefix(fileName: String): String = {
+        val driveLetter = s"^[A-Za-z]:".r
+        val supportDriveLetter =
+          build.Platform.isWindows || build.Platform.isCygwin || build.Platform.isMsys
+        // using "/" instead of `File.separator`
+        // we get "/C:a/b/c.txt" even in windows for some reasons.
+        val noDriveLetter =
+          driveLetter.findFirstIn(fileName.stripPrefix("/")) match {
```

Comment:
I thought a bit about this change recently and the fact that we need to have so much special logic for paths handling gives me idea, that maybe that's not the best solution for this problem. Recently I though about approach which might be simpler, easier to maintain and would not depend from different paths representations/file systems. 

So we had original implementation having 1 file per package. Now we have 1 file per source directory, but requires lots of dedicated logic when compared to original solution. 
What if we would simply made 2 group by operations: 
1. group by package name, to get a fine granularity of files 
2. group package definitions by directory, add index suffix to distinguish files defined in different compilation units

Typically most of the files would be contained within the same source directory, because that's how we typically structure Scala/JVM projects. If some definitions would be in the same package, but would be defined in different directories, then probably we are dealing with 2 different compilation units (eg. code coming from different libraries using the same package name). 

If we'd only split paths by parent directory and then optionally would add suffix to file name,  then we won't need to have this special logic.

Reply:
Thanks @WojciechMazur I agree that the complication is caused because we try to use the directory name for `.ll` files. Naming the `.ll` files based on package names would be easier.

I'm wondering which do you mean (1) or (2) or I'm misunderstanding, could you elaborate on it?

- (1) Do you mean "Split `Defn`s by packages into groups, and then combines the groups if they are defined in the same Scala directory (or file)"?
  - If this is the case, it seems like this is equivalent to "Split `Defn`s into groups by directories" and name them based on package name
  - but how should we name if a directory has multiple packages? I'm not sure how suffix (to `.ll` files, right?) solve the naming issue.
- (2) Or do you mean "Split `Defn`s by packages into groups, and then split again by directory"?
  - In this case, adding suffix to `package_name.ll` if multiple Scala directories defines the package.
  - However, it doesn't work since we would have multiple `.ll` files per packages even if those packages are defined in the same file.

For (1), do you think it's ok to name the `.ll` files just like
- `1.ll`, `2.ll`, ... 
- `pkg1.ll` or `pkg1pkg2.ll` if the directory defines multiple packages (`pkg1` and `pkg2`)

Reply:
I mean like https://github.com/scala-native/scala-native/pull/3466/commits/3ca24e6b0240e8611e080e18ac9fdd200fc8a56a (which doesn't work properly yet 🤔 )

```
❯ ls unit-tests/native/.3/target/**/*.ll | grep _
unit-tests/native/.3/target/scala-3.1.3/native-test/__empty_package.ll
unit-tests/native/.3/target/scala-3.1.3/native-test/__empty_package_java.lang_scala.scalanative.runtime.ll
unit-tests/native/.3/target/scala-3.1.3/native-test/com.novocode.junit_com.novocode.junit.JUnitFrameworkL33com.novocode.junit.ll
unit-tests/native/.3/target/scala-3.1.3/native-test/org.scalanative.testsuite.javalib.util.concurrent_org.scalanative.testsuite.javalib.util.concurrent.atomic.ll
unit-tests/native/.3/target/scala-3.1.3/native-test/org.scalanative.testsuite.posixlib_org.scalanative.testsuite.posixlib.sys.ll
unit-tests/native/.3/target/scala-3.1.3/native-test/org.scalanative.testsuite.utils_scala.scalanative.junit.utils.ll
unit-tests/native/.3/target/scala-3.1.3/native-test/scala.scalanative.annotation_scala.scalanative.annotation.AlignTestCases.ll
unit-tests/native/.3/target/scala-3.1.3/native-test/scala.scalanative.junit_scala.scalanative.junit.JUnitFrameworkL38scala.scalanative.junit.ll
```

Reply:
Maybe it's good to stick with partitioning by parent directory even though we need some special treatment for Windows?

Reply:
It makes the `.ll` file names no sense, but using the hash of the directory would be the easiest

```diff
--- a/tools/src/main/scala/scala/scalanative/codegen/llvm/CodeGen.scala
+++ b/tools/src/main/scala/scala/scalanative/codegen/llvm/CodeGen.scala
@@ -12,6 +12,7 @@ import scala.scalanative.build
 import scala.scalanative.linker.ReachabilityAnalysis
 import scala.scalanative.util.{Scope, partitionBy, procs}
 import java.nio.file.StandardCopyOption
+import java.security.MessageDigest

 import scala.scalanative.build.ScalaNative
 import scala.scalanative.codegen.{Metadata => CodeGenMetadata}
@@ -131,26 +132,27 @@ object CodeGen {
         // To address this issue, we partition into multiple LLVM IR files per Scala source file originated from.
         // This will ensure that each LLVM IR file only references a single Scala source file,
         // which will prevent the Darwin linker failing to generate N_OSO symbols.
+        val md5 = MessageDigest.getInstance("MD5")
         Future
         // which will prevent the Darwin linker failing to generate N_O
SO symbols.
+        val md5 = MessageDigest.getInstance("MD5")
         Future
           .traverse(assembly.groupBy(x => sourceDirOf(x.pos)).toSeq) {
             case (dir, defns) =>
               Future {
-                val path = dropPrefix(dir.stripSuffix("/"))
-                val outFile = config.workDir.resolve(s"$path.ll")
+                val hash = md5.digest(dir.getBytes).toString
+                val outFile = config.workDir.resolve(s"$hash.ll")
                 val ownerDirectory = outFile.getParent()

-                ctx.addEntry(path, defns)
-                if (ctx.shouldCompile(path)) {
+                ctx.addEntry(hash, defns)
+                if (ctx.shouldCompile(hash)) {
                   val sorted = defns.sortBy(_.name)
                   if (!Files.exists(ownerDirectory))
                     Files.createDirectories(ownerDirectory)
-                  Impl(env, sorted).gen(path, workDir)
+                  Impl(env, sorted).gen(hash, workDir)
                 } else {
                   assert(ownerDirectory.toFile.exists())
                   config.logger.debug(
-                    s"Content of package has not changed, skiping gener
ation of $path.ll"
+                    s"Content of package has not changed, skiping gener
ation of $hash.ll"
                   )
-                  config.workDir.resolve(s"$path.ll")
+                  config.workDir.resolve(s"$hash.ll")
                 }
               }
           }
```

Reply:
Talked with @WojciechMazur and consent that we're going to group `Defn`s by parent directory and name the `.ll` files based on the hash of parent directory like https://github.com/scala-native/scala-native/pull/3466#discussion_r1339538196. This will fix the original issue + don't need special handling for platform specific paths.
We'd been naming the `.ll` files based on the package names so we can easily find the LLVM IR that corresponds to the package and debug LLVM IR files. However, in reality we usually find the LLVM IR files based on `grep` instead of Scala package, so hopefully it isn't a problem.